File: mnemonic-hadoop/mnemonic-hadoop-mapreduce/src/main/java/org/apache/mnemonic/hadoop/MneDurableInputSession.java
/**
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package org.apache.mnemonic.hadoop;

import java.nio.file.Files;
import java.nio.file.LinkOption;
import java.nio.file.Paths;
import java.util.ArrayList;
import java.util.Iterator;
import java.util.List;

import org.apache.hadoop.conf.Configuration;
import org.apache.hadoop.fs.Path;
import org.apache.hadoop.mapreduce.TaskAttemptContext;
import org.apache.mnemonic.ConfigurationException;
import org.apache.mnemonic.DurableType;
import org.apache.mnemonic.NonVolatileMemAllocator;
import org.apache.mnemonic.Utils;
import org.apache.mnemonic.collections.DurableSinglyLinkedList;
import org.apache.mnemonic.collections.DurableSinglyLinkedListFactory;
import org.apache.mnemonic.sessions.DurableInputSession;
import org.apache.mnemonic.sessions.SessionIterator;

/**
 * This class extends DurableInputSession and is responsible for handling 
 * non-volatile memory-based input sessions using Hadoop’s TaskAttemptContext.
 * 
 * @param <V> The type of elements stored in the session (must be durable).
 */
public class MneDurableInputSession<V>
    extends DurableInputSession<V, NonVolatileMemAllocator, Void, Void> {

  private TaskAttemptContext taskAttemptContext;
  private Configuration configuration;
  private Iterator<String> m_fp_iter;

  /**
   * Constructor initializes the session using either a TaskAttemptContext or
   * a standalone Configuration object.
   * 
   * @param taskAttemptContext Task-specific context (optional).
   * @param configuration      Hadoop Configuration.
   * @param paths              Array of file paths for input.
   * @param prefix             Configuration prefix for memory service.
   */
  public MneDurableInputSession(TaskAttemptContext taskAttemptContext,
      Configuration configuration, Path[] paths, String prefix) {
    // Check if both the TaskAttemptContext and Configuration are null
    if (null == taskAttemptContext && null == configuration) {
      throw new ConfigurationException("Session is not configured properly");
    }
    
    // If taskAttemptContext is provided, use it to set configuration
    if (null != taskAttemptContext) {
      setTaskAttemptContext(taskAttemptContext);
      setConfiguration(taskAttemptContext.getConfiguration());
    } else {
      setConfiguration(configuration);
    }
    
    // Initialize file paths and configuration
    initialize(paths, prefix);
  }

  /**
   * Initializes file paths and reads the configuration prefix for the session.
   * 
   * @param paths  Array of file paths.
   * @param prefix Configuration prefix for memory service.
   */
  public void initialize(Path[] paths, String prefix) {
    List<String> fpathlist = new ArrayList<String>();
    
    // Validate each path to ensure it's a regular file
    for (Path p : paths) {
      if (!Files.isRegularFile(Paths.get(p.toString()), LinkOption.NOFOLLOW_LINKS)) {
        throw new UnsupportedOperationException();
      }
      fpathlist.add(p.toString());
    }
    
    // Set up an iterator for the file paths
    m_fp_iter = fpathlist.iterator();
    
    // Read the configuration
    readConfig(prefix);
  }

  /**
   * Validates the configuration for the session, checking if durable types 
   * and entity factory proxies are properly set.
   */
  public void validateConfig() {
    if (getDurableTypes().length < 1) {
      throw new ConfigurationException("The durable type of record parameters does not exist");
    } else {
      // Check if the first durable type is a durable entity and if proxies are set
      if (DurableType.DURABLE == getDurableTypes()[0]
          && getEntityFactoryProxies().length < 1) {
        throw new ConfigurationException("The durable entity proxy of record parameters does not exist");
      }
    }
  }

  /**
   * Reads the configuration using a prefix to set service name, durable types,
   * entity factory proxies, and slot key ID for memory allocation.
   * 
   * @param prefix Configuration prefix.
   */
  public void readConfig(String prefix) {
    Configuration conf = getConfiguration();
    
    // Check if configuration is set
    if (conf == null) {
      throw new ConfigurationException("Configuration has not yet been set");
    }
    
    // Set configuration values for memory service, durable types, entity proxies, and slot key ID
    setServiceName(MneConfigHelper.getMemServiceName(conf, MneConfigHelper.DEFAULT_INPUT_CONFIG_PREFIX));
    setDurableTypes(MneConfigHelper.getDurableTypes(conf, MneConfigHelper.DEFAULT_INPUT_CONFIG_PREFIX));
    setEntityFactoryProxies(Utils.instantiateEntityFactoryProxies(
        MneConfigHelper.getEntityFactoryProxies(conf, MneConfigHelper.DEFAULT_INPUT_CONFIG_PREFIX)));
    setSlotKeyId(MneConfigHelper.getSlotKeyId(conf, MneConfigHelper.DEFAULT_INPUT_CONFIG_PREFIX));
    
    // Validate the configuration to ensure it's correct
    validateConfig();
  }

  /**
   * Initialize the session iterator. In this case, it's a simple success check.
   * 
   * @param sessiter The session iterator to initialize.
   * @return Always returns true since no additional work is required for init.
   */
  @Override
  protected boolean init(SessionIterator<V, NonVolatileMemAllocator, Void, Void> sessiter) {
    return true;
  }

  /**
   * Prepares the next memory pool for the session and restores data if available.
   * 
   * @param sessiter The session iterator used to restore and iterate over durable lists.
   * @return True if initialization is successful, false otherwise.
   */
  @Override
  protected boolean initNextPool(SessionIterator<V, NonVolatileMemAllocator, Void, Void> sessiter) {
    boolean ret = false;
    
    // Close and reset the allocator if already initialized
    if (sessiter.getAllocator() != null) {
      sessiter.getAllocator().close();
      sessiter.setAllocator(null);
    }
    
    // If there are more files to process
    if (null != m_fp_iter && m_fp_iter.hasNext()) {
      // Create a new memory allocator with the next file path
      sessiter.setAllocator(new NonVolatileMemAllocator(Utils.getNonVolatileMemoryAllocatorService(
          getServiceName()), 1024000L, m_fp_iter.next(), false));
      
      // Check if the allocator was successfully created
      if (null != sessiter.getAllocator()) {
        sessiter.setHandler(sessiter.getAllocator().getHandler(getSlotKeyId()));
        
        // If the handler is valid, restore the durable singly linked list
        if (0L != sessiter.getHandler()) {
          DurableSinglyLinkedList<V> dsllist = DurableSinglyLinkedListFactory.restore(
              sessiter.getAllocator(), getEntityFactoryProxies(), getDurableTypes(), sessiter.getHandler(), false);
          
          // Set the iterator if the list was restored successfully
          if (null != dsllist) {
            sessiter.setIterator(dsllist.iterator());
            ret = null != sessiter.getIterator();
          }
        }
      }
    }
    return ret;
  }

  // Getter for TaskAttemptContext
  public TaskAttemptContext getTaskAttemptContext() {
    return taskAttemptContext;
  }

  // Setter for TaskAttemptContext
  public void setTaskAttemptContext(TaskAttemptContext taskAttemptContext) {
    this.taskAttemptContext = taskAttemptContext;
  }

  // Getter for Configuration
  public Configuration getConfiguration() {
    return configuration;
  }

  // Setter for Configuration
  public void setConfiguration(Configuration configuration) {
    this.configuration = configuration;
  }
}

File: mnemonic-spark/mnemonic-spark-core/src/main/scala/org/apache/mnemonic/spark/MneDurableInputSession.scala
/*
 * Licensed to the Apache Software Foundation (ASF) under one or more
 * contributor license agreements.  See the NOTICE file distributed with
 * this work for additional information regarding copyright ownership.
 * The ASF licenses this file to You under the Apache License, Version 2.0
 * (the "License"); you may not use this file except in compliance with
 * the License.  You may obtain a copy of the License at
 *
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package org.apache.mnemonic.spark

import java.io.{File, FileOutputStream}
import java.util.concurrent.TimeUnit

import scala.reflect.ClassTag
import scala.collection.JavaConverters._
import org.apache.mnemonic.ConfigurationException
import org.apache.mnemonic.DurableType
import org.apache.mnemonic.EntityFactoryProxy
import org.apache.mnemonic.NonVolatileMemAllocator
import org.apache.mnemonic.Utils
import org.apache.mnemonic.collections.DurableSinglyLinkedList
import org.apache.mnemonic.collections.DurableSinglyLinkedListFactory
import org.apache.mnemonic.sessions.{DurableInputSession, ObjectCreator, SessionIterator, TransformFunction}

case class SessionIteratorState(var position: Int, var streamMode: Option[Boolean])

private[spark] class MneDurableInputSession[V: ClassTag, T: ClassTag] (
    serviceName: String,
    durableTypes: Array[DurableType],
    entityFactoryProxies: Array[EntityFactoryProxy],
    slotKeyId: Long,
    var memPoolListGen: ()=>Option[Array[File]],
    outsess: MneDurableOutputSession[V], /*would be closed once no more element from source iterator*/
    srciterator: Iterator[T],
    f: (T, ObjectCreator[V, NonVolatileMemAllocator]) => Option[V],
    claimfile: File)
    extends DurableInputSession[V, NonVolatileMemAllocator, T, SessionIteratorState] {


  var iteratorHolder: SessionIterator[V, NonVolatileMemAllocator, T, SessionIteratorState] = null

  var streamMode: Option[Boolean] = None
  initialize

  def initialize: Unit = {
    setServiceName(serviceName)
    setDurableTypes(durableTypes)
    setEntityFactoryProxies(entityFactoryProxies)
    setSlotKeyId(slotKeyId)
    setOutSession(outsess)
    setTransformFunction(new TransformFunction[V, NonVolatileMemAllocator, T] {
      override def transform(value:T, objcreator:ObjectCreator[V, NonVolatileMemAllocator]):V = {
         f(value, objcreator).getOrElse(null).asInstanceOf[V]
      }
    })
  }

  def updateMode: Option[Boolean] = {
    if (null != srciterator && null != getOutSession &&
        null != getTransformFunction && null != claimfile) {
      streamMode = Option(true)
    } else if (null != memPoolListGen) {
      streamMode = Option(false)
    } else {
      throw new ConfigurationException("Cannot determine which mode Input Session works on")
    }
    streamMode
  }

  def transformAhead(): Unit = {
    if (updateMode.get) {
      streamMode = Option(tryClaim)
    }
    if (streamMode.get) {
      try {
        for (item <- srciterator) {
          f(item, outsess) match {
            case Some(res) => outsess.post(res)
            case None =>
          }
        }
      } finally {
        outsess.close
        setOutSession(null)
        setTransformFunction(null)
      }
      streamMode = Option(false)
    }
  }

  override def init(sessiter: SessionIterator[V, NonVolatileMemAllocator, T, SessionIteratorState]): Boolean = {
    sessiter.setState(SessionIteratorState(0, streamMode))
    null != sessiter.getState
  }

  def waitNextPool(file: File): Unit = {
    val lckfile: File = new File(file.toString + ".lck")
    while (lckfile.exists) {
      TimeUnit.SECONDS.sleep(1)
    }
  }

  def tryClaim: Boolean = {
    if (null == claimfile) {
      false
    } else {
      if (claimfile.exists) {
        false
      } else {
        new FileOutputStream(claimfile).close
        true
      }
    }
  }

  override def initNextPool(sessiter: SessionIterator[V, NonVolatileMemAllocator, T, SessionIteratorState]): Boolean = {
    var ret: Boolean = false
    var flist: Array[File] = null
    val state: SessionIteratorState = sessiter.getState

    if (null != sessiter.getAllocator) {
      sessiter.getAllocator.close
      sessiter.setAllocator(null)
    }

    flist = memPoolListGen().getOrElse(null)
    if (state.streamMode.isEmpty) {
      state.streamMode = Option(tryClaim)
      if (state.streamMode.get) {
        println(s"Input Session run in stream mode with ${claimfile}")
      } else {
        println(s"Input Session run in durable mode with ${claimfile}")
      }
    }

    if (!state.streamMode.get) {
      if (null != flist && flist.length > state.position) {
        waitNextPool(flist(state.position))
        sessiter.setAllocator(new NonVolatileMemAllocator(Utils.getNonVolatileMemoryAllocatorService(
          getServiceName), 1024000L, flist(state.position).toString, false))
        assert(null != sessiter.getAllocator)
        sessiter.setHandler(sessiter.getAllocator.getHandler(getSlotKeyId))
        if (0L != sessiter.getHandler) {
          val dsllist: DurableSinglyLinkedList[V] = DurableSinglyLinkedListFactory.restore(
            sessiter.getAllocator, getEntityFactoryProxies, getDurableTypes, sessiter.getHandler, false)
          if (null != dsllist) {
            sessiter.setIterator(dsllist.iterator)
            ret = null != sessiter.getIterator
            if (ret) {
              println(s"Input Session obtained the iterator of ${flist(state.position)}")
            }
          } else {
            throw new ConfigurationException("The singly linked list cannot be restored")
          }
        } else {
          throw new ConfigurationException("The value of slot handler is null")
        }
        state.position += 1
      }
    } else {
      sessiter.setSourceIterator(srciterator.asJava)
      ret = null != sessiter.getSourceIterator
      if (ret) {
        println(s"Input Session obtained the iterator of upstream with ${claimfile}")
      }
    }
    ret
  }

  override def iterator: SessionIterator[V, NonVolatileMemAllocator, T, SessionIteratorState] = {
    if (null != iteratorHolder) {
      iteratorHolder.close
      println("force to close an input session iterator prematurely")
    }
    iteratorHolder = super.iterator
    iteratorHolder
  }

  override def close: Unit = {
    super.close();
    if (null != getOutSession) {
      getOutSession.close
      setOutSession(null)
    }
    if (null != iteratorHolder) {
      iteratorHolder.close
      iteratorHolder = null
    }
  }

}

object MneDurableInputSession {
  def apply[V: ClassTag](
                          serviceName: String,
                          durableTypes: Array[DurableType],
                          entityFactoryProxies: Array[EntityFactoryProxy],
                          slotKeyId: Long,
                          memPoolListGen: ()=>Option[Array[File]]): MneDurableInputSession[V, Nothing] = {
    val ret = new MneDurableInputSession[V, Nothing] (
        serviceName, durableTypes, entityFactoryProxies,
        slotKeyId, memPoolListGen, null, null, null, null)
    ret
  }
  def apply[V: ClassTag, T: ClassTag](
                          serviceName: String,
                          durableTypes: Array[DurableType],
                          entityFactoryProxies: Array[EntityFactoryProxy],
                          slotKeyId: Long,
                          memPoolListGen: ()=>Option[Array[File]],
                          outsess: MneDurableOutputSession[V],
                          srciterator: Iterator[T],
                          f: (T, ObjectCreator[V, NonVolatileMemAllocator]) => Option[V],
                          claimfile: File):
                          MneDurableInputSession[V, T] = {
    val ret = new MneDurableInputSession[V, T] (
      serviceName, durableTypes, entityFactoryProxies,
      slotKeyId, memPoolListGen, outsess, srciterator, f, claimfile)
    ret
  }
}
